botocore/client.py

```diff
@@ -199,37 +176,59 @@ def _resolve_verify_value(self, provided_verify_value, scoped_config):
         elif scoped_config is not None:
             return scoped_config.get('ca_bundle')
 
+    def _determine_region_name(self, region_name, client_config):
+        # Determine what region the user provided either via the
+        # region_name argument or the client_config. Note that
+        # endpoint resolvers are responsible for enforcing whether
+        # or not a region is required.
+        if region_name is None:
+            if client_config and client_config.region_name is not None:
+                region_name = client_config.region_name
+        return region_name
+
+    def _resolve_endpoint(self, region_name, endpoint_config, endpoint_url,
```

Comment:
Sounds good. I feel like it was worth mentioning as the logic in this class seems to keep growing and growing and at some point I feel like it will be due for a refactoring before it ends up like endpoint where too much unneeded logic lives directly in the class.